sdk/cosmos/azure_data_cosmos/src/models/patch_operations.rs

```diff
@@ -0,0 +1,323 @@
+    ///
+    /// See the [type documentation](PatchDocument) for more information on patch operations.
+    ///
+    /// # Arguments
+    /// * `path` - The path to the property to increment.
+    /// * `value` - The amount to increment by. Integers can be specified directly in this parameter. Use [`serde_json::Number::from_f64`] to create a floating-point number.
+    pub fn with_increment(
+        mut self,
+        path: String,
+        value: serde_json::Number,
+    ) -> Result<Self, serde_json::Error> {
+        self.operations
+            .push(PatchOperation::Increment { path, value });
+        Ok(self)
+    }
+
+    /// Adds a new "remove" operation to the patch document.
+    ///
+    /// See the [type documentation](PatchDocument) for more information on patch operations.
+    ///
+    /// # Arguments
+    /// * `path` - The path to the property to remove.
+    pub fn with_remove(mut self, path: String) -> Result<Self, serde_json::Error> {
+        self.operations.push(PatchOperation::Remove { path });
+        Ok(self)
+    }
+
+    /// Adds a new "replace" operation to the patch document.
+    ///
+    /// See the [type documentation](PatchDocument) for more information on patch operations.
+    ///
+    /// # Arguments
+    /// * `path` - The path to the property to remove.
+    /// * `value` - The value to replace the property with.
+    pub fn with_replace(
+        mut self,
+        path: String,
+        value: impl Serialize,
+    ) -> Result<Self, serde_json::Error> {
+        self.operations.push(PatchOperation::Replace {
+            path,
+            value: serde_json::to_value(value)?,
+        });
+        Ok(self)
+    }
+
+    /// Adds a new "set" operation to the patch document.
+    ///
+    /// See the [type documentation](PatchDocument) for more information on patch operations.
+    ///
+    /// # Arguments
+    /// * `path` - The path to the property to remove.
+    /// * `value` - The value to set the property to.
+    pub fn with_set(
+        mut self,
+        path: String,
+        value: impl Serialize,
+    ) -> Result<Self, serde_json::Error> {
+        self.operations.push(PatchOperation::Set {
+            path,
+            value: serde_json::to_value(value)?,
+        });
+        Ok(self)
+    }
+
+    /// Adds a new "move" operation to the patch document.
+    ///
+    /// See the [type documentation](PatchDocument) for more information on patch operations.
+    ///
+    /// # Arguments
+    /// * `from` - The path to the property to move.
+    /// * `to` - The path to move the property to.
+    pub fn with_move(mut self, from: String, to: String) -> Result<Self, serde_json::Error> {
+        self.operations.push(PatchOperation::Move { from, to });
+        Ok(self)
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
+#[serde(tag = "op")]
+#[serde(rename_all = "camelCase")]
+pub enum PatchOperation {
+    Add {
+        path: String,
+        value: serde_json::Value,
```

Comment:
Storing this as `serde_json::Value` forces us to "pre-serialize" the value to JSON and then "serialize" it again from `serde_json::Value` in to the final body. However, our options are limited here, because `Serialize` isn't object-safe. 

There is the [`erased_serde` crate](https://crates.io/crates/erased-serde), which provides a type-erased `Serialize` that **is** object-safe, and we could use that and store `Box<dyn erased_serde::Serialize>` here.

This also means we're exposing `serde_json::Value` to our public API surface, which isn't ideal, but may be somewhat unavoidable here unless we use a newtype.

Reply:
I'd rather not take yet another crate (`erased_serde`) just for this. While not ideal to expose `serde_json` types, I'm not as worried about the serde family of crates. They're ubiquitous. I'm also not too concerned about object safety unless we know we need it somewhere. So many things we do or even use are not object safe that it feels like a bygone conclusion.

That said, I do like the newtype idea here and for `Number`, if possible, as well. Also gives us some flexibility to impl other traits or add methods as needed so we'll own the types.